bug: TLS credentials not passed to tiflow in old architecture mode

### Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. **Set up a TLS-enabled TiDB cluster** with self-signed certificates:

```bash
# Generate test certificates
mkdir -p /tmp/cdc-test/certs && cd /tmp/cdc-test/certs
openssl genrsa -out ca.key 2048
openssl req -new -x509 -key ca.key -out ca.crt -days 1 -subj "/CN=test-ca"
openssl genrsa -out server.key 2048
openssl req -new -key server.key -out server.csr -subj "/CN=test-server"
openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 1
```

2. **Start TiDB cluster with TLS** (PD listening on HTTPS)

3. **Build TiCDC** from master:
```bash
git clone https://github.com/pingcap/ticdc.git
cd ticdc
make build
```

4. **Start TiCDC with TLS credentials**:
```bash
./bin/cdc server --pd=https://127.0.0.1:2379 \
  --ca=/tmp/cdc-test/certs/ca.crt \
  --cert=/tmp/cdc-test/certs/server.crt \
  --key=/tmp/cdc-test/certs/server.key \
  --log-level=debug
```

### 2. What did you expect to see? (Required)

TiCDC should connect to PD using HTTPS with the provided TLS credentials:
```
[INFO] [client.go:291] ["[pd] create pd client with endpoints"] [pd-address="[https://127.0.0.1:2379]"]
```

And the security config should show the certificate paths:
```
"security":{"ca-path":"/tmp/cdc-test/certs/ca.crt","cert-path":"/tmp/cdc-test/certs/server.crt","key-path":"/tmp/cdc-test/certs/server.key",...}
```

### 3. What did you see instead (Required)

TiCDC connects to PD using HTTP, ignoring TLS credentials:
```
[INFO] [client.go:291] ["[pd] create pd client with endpoints"] [pd-address="[http://127.0.0.1:2379]"]
```

The security config shows empty paths despite CLI flags being passed:
```
"security":{"ca-path":"","cert-path":"","key-path":"","cert-allowed-cn":null,...}
```

This causes connection failures when PD requires TLS:
```
[ERROR] [client_interceptor.go:165] ["grpc request error"] [error="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp 127.0.0.1:2379: connect: connection refused\""]
```

### 4. What is your TiCDC version? (Required)

Current master branch (any version with new architecture code that can fall back to old architecture mode)

### 5. What is your Downstream Database and version? (mysql/tidb/kafka)

Any downstream - this is a server startup issue

### 6. Additional Information (Paste cdc_stderr.log/cdc.log here or provide link if available)

**Root Cause:**

In old architecture mode, `NewCmdServer()` calls `runTiFlowServer()` directly. The `runTiFlowServer()` function:
1. Marshals `o.serverConfig` to JSON
2. Unmarshals into tiflow's `ServerConfig`
3. Passes `o.caPath`, `o.certPath`, `o.keyPath` to `oldOptions`

The problem is that `o.serverConfig.Security` is never populated from the CLI flags before JSON marshaling. While `oldOptions.CaPath/CertPath/KeyPath` are set, the tiflow `Run()` function calls `complete()` which copies from `o.ServerConfig.Security` (the unmarshaled empty struct) back to `cfg.Security`, overwriting any credentials.

See tiflow's `complete()` at lines 191-198 in `pkg/cmd/server/server.go`:
```go
case "ca":
    cfg.Security.CAPath = o.ServerConfig.Security.CAPath  // Empty!
```

**Proposed Fix:**

Add `o.serverConfig.Security = o.getCredential()` in `runTiFlowServer()` before JSON marshaling:

```go
func runTiFlowServer(o *options, cmd *cobra.Command) error {
    // Populate security credentials from CLI flags before marshaling to JSON.
    o.serverConfig.Security = o.getCredential()
    
    cfgData, err := json.Marshal(o.serverConfig)
    // ... rest of function
}
```

**Impact:**
- **Severity**: High - TiCDC cannot connect to TLS-enabled clusters in old architecture mode
- **Scope**: Any TiCDC deployment using TLS with old architecture mode

--- a/cmd/cdc/server/server.go
+++ b/cmd/cdc/server/server.go
@@ -297,6 +297,12 @@ func isNewArchEnabled(o *options) bool {
 }
 
 func runTiFlowServer(o *options, cmd *cobra.Command) error {
+	// Populate security credentials from CLI flags before marshaling to JSON.
+	// In old architecture mode, complete() is not called, so we need to transfer
+	// TLS credentials (--ca, --cert, --key) to serverConfig.Security here.
+	// Without this, the Security struct is empty and tiflow uses HTTP instead of HTTPS.
+	o.serverConfig.Security = o.getCredential()
+
 	cfgData, err := json.Marshal(o.serverConfig)
 	if err != nil {
 		return errors.Trace(err)
@@ -312,10 +318,6 @@ func runTiFlowServer(o *options, cmd *cobra.Command) error {
 	oldOptions.ServerConfig = &oldCfg
 	oldOptions.ServerPdAddr = strings.Join(o.pdEndpoints, ",")
 	oldOptions.ServerConfigFilePath = o.serverConfigFilePath
-	oldOptions.CaPath = o.caPath
-	oldOptions.CertPath = o.certPath
-	oldOptions.KeyPath = o.keyPath
-	oldOptions.AllowedCertCN = o.allowedCertCN
 
 	return tiflowServer.Run(&oldOptions, cmd)
 }

--- a/cmd/cdc/server/server_test.go
+++ b/cmd/cdc/server/server_test.go
@@ -1,0 +1,129 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/json"
+	"testing"
+
+	tiflowConfig "github.com/pingcap/tiflow/pkg/config"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRunTiFlowServerPopulatesSecurityConfig(t *testing.T) {
+	// This test verifies that TLS credentials from CLI flags are properly
+	// transferred to serverConfig.Security before being passed to tiflow.
+	// See https://github.com/pingcap/ticdc/issues/3718
+
+	o := newOptions()
+	o.caPath = "/path/to/ca.crt"
+	o.certPath = "/path/to/server.crt"
+	o.keyPath = "/path/to/server.key"
+	o.allowedCertCN = "cn1,cn2"
+
+	// Verify that before calling getCredential, serverConfig.Security is empty
+	require.Empty(t, o.serverConfig.Security.CAPath)
+	require.Empty(t, o.serverConfig.Security.CertPath)
+	require.Empty(t, o.serverConfig.Security.KeyPath)
+
+	// Simulate what runTiFlowServer does: populate Security from CLI flags
+	o.serverConfig.Security = o.getCredential()
+
+	// Verify Security is now populated
+	require.Equal(t, "/path/to/ca.crt", o.serverConfig.Security.CAPath)
+	require.Equal(t, "/path/to/server.crt", o.serverConfig.Security.CertPath)
+	require.Equal(t, "/path/to/server.key", o.serverConfig.Security.KeyPath)
+	require.Equal(t, []string{"cn1", "cn2"}, o.serverConfig.Security.CertAllowedCN)
+
+	// Verify that JSON marshaling preserves the Security config
+	cfgData, err := json.Marshal(o.serverConfig)
+	require.NoError(t, err)
+
+	var oldCfg tiflowConfig.ServerConfig
+	err = json.Unmarshal(cfgData, &oldCfg)
+	require.NoError(t, err)
+
+	// This is the critical assertion: tiflow's ServerConfig.Security
+	// should have the TLS credentials after unmarshaling
+	require.Equal(t, "/path/to/ca.crt", oldCfg.Security.CAPath)
+	require.Equal(t, "/path/to/server.crt", oldCfg.Security.CertPath)
+	require.Equal(t, "/path/to/server.key", oldCfg.Security.KeyPath)
+	require.Equal(t, []string{"cn1", "cn2"}, oldCfg.Security.CertAllowedCN)
+}
+
+func TestGetCredential(t *testing.T) {
+	testCases := []struct {
+		name           string
+		caPath         string
+		certPath       string
+		keyPath        string
+		allowedCertCN  string
+		expectedCAPath string
+		expectedCert   string
+		expectedKey    string
+		expectedCertCN []string
+	}{
+		{
+			name:           "all fields populated",
+			caPath:         "/ca/path",
+			certPath:       "/cert/path",
+			keyPath:        "/key/path",
+			allowedCertCN:  "test-cn",
+			expectedCAPath: "/ca/path",
+			expectedCert:   "/cert/path",
+			expectedKey:    "/key/path",
+			expectedCertCN: []string{"test-cn"},
+		},
+		{
+			name:           "multiple CNs",
+			allowedCertCN:  "cn1,cn2,cn3",
+			expectedCertCN: []string{"cn1", "cn2", "cn3"},
+		},
+		{
+			name:           "empty CN",
+			allowedCertCN:  "",
+			expectedCertCN: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			o := newOptions()
+			o.caPath = tc.caPath
+			o.certPath = tc.certPath
+			o.keyPath = tc.keyPath
+			o.allowedCertCN = tc.allowedCertCN
+
+			cred := o.getCredential()
+
+			require.Equal(t, tc.expectedCAPath, cred.CAPath)
+			require.Equal(t, tc.expectedCert, cred.CertPath)
+			require.Equal(t, tc.expectedKey, cred.KeyPath)
+			require.Equal(t, tc.expectedCertCN, cred.CertAllowedCN)
+		})
+	}
+}
+
+func TestNewOptionsDefaultSecurity(t *testing.T) {
+	o := newOptions()
+
+	// serverConfig should be initialized with default values
+	require.NotNil(t, o.serverConfig)
+	require.NotNil(t, o.serverConfig.Security)
+
+	// Security should have empty paths by default
+	require.Empty(t, o.serverConfig.Security.CAPath)
+	require.Empty(t, o.serverConfig.Security.CertPath)
+	require.Empty(t, o.serverConfig.Security.KeyPath)
+}

